File: Leetcode/leetcode-0930-BinarySubarraysWithSum-[array]-[medium]/Solution.java
class Solution {
    public int numSubarraysWithSum(int[] nums, int goal) {
        int[] prefix = new int[nums.length];
        prefix[0] = nums[0];
        for (int i = 1; i < nums.length; i++) {
            prefix[i] = prefix[i - 1] + nums[i];
        }

        HashMap<Integer, Integer> prefixHistory = new HashMap<>();
        prefixHistory.put(0, 1);
        int cnt = 0;
        for (int i = 0; i < nums.length; i++) {
            if (prefixHistory.get(prefix[i] - goal) != null) {
                cnt += prefixHistory.getOrDefault(prefix[i] - goal, 0);
            }

            prefixHistory.put(prefix[i], prefixHistory.getOrDefault(prefix[i], 0) + 1);
        }
        

        return cnt;
    }
}

File: Leetcode/leetcode-0930-BinarySubarraysWithSum-[array]-[medium]/Solution.py
class Solution:
    def numSubarraysWithSum(self, nums: List[int], goal: int) -> int:
        cnt = 0
        prefix = 0
        prefixHistory = dict({})
        prefixHistory.update({ 0: 1 })

        for num in nums:
            prefix += num
            if (prefixHistory.get(prefix - goal) != None):
                cnt += prefixHistory.get(prefix - goal, 0)

            prefixHistory.update({ prefix: prefixHistory.get(prefix, 0) + 1 })

        return cnt